src/NumericalAlgorithms/Interpolation/LinearLeastSquares.cpp

```diff
@@ -0,0 +1,80 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "NumericalAlgorithms/Interpolation/LinearLeastSquares.hpp"
+
+#include <algorithm>
+#include <cstddef>
+#include <gsl/gsl_spline.h>
+#include <memory>
+#include <pup.h>
+#include <pup_stl.h>
+#include <vector>
+
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/Gsl.hpp"
+
+#include <iostream>
+
+namespace intrp {
+template <size_t Order>
+LinearLeastSquares<Order>::LinearLeastSquares(const size_t num_observations)
+    : num_observations_(num_observations) {}
+
+template <size_t Order>
+double LinearLeastSquares<Order>::interpolate(
+    const std::array<double, Order + 1> coefficients,
+    const double x_to_interp_to) {
+  double result = 0;
+  for (size_t i = 0; i < Order + 1; i++) {
+    result += gsl::at(coefficients, i) * pow(x_to_interp_to, i);
+  }
+  return result;
+}
+
+template <size_t Order>
+std::array<double, Order + 1> LinearLeastSquares<Order>::fit_coefficients(
+    const std::vector<double> x_values, const std::vector<double> y_values) {
+  std::array<double, Order + 1> coefficients{{}};
+  X = gsl_matrix_alloc(num_observations_, Order + 1);
```

Comment:
You're still doing a lot of memory allocations here. Can you move these memory allocations into the constructor and then the `gsl_*_free()` calls below into the destructor? Maybe the `gsl_multifit_linear_alloc`